request_test.go

```diff
@@ -1601,3 +1602,634 @@ func TestRequestErrorConflictMultipart(t *testing.T) {
+						assert.Nil(t, r.GetBody)
+					} else {
+						t.Fatalf("invalid request URL")
+					}
+
+					return true
+				}).WithMaxRedirect(1)
+
+			client := &http.Client{
+				Transport: tp,
+			}
+
+			config := Config{
+				Client:   client,
+				Reporter: reporter,
+			}
+
+			req := NewRequest(config, http.MethodPut, "/url").
+				WithRedirectPolicy(FollowAllRedirects).
+				WithMaxRedirects(1)
+			req.chain.assertOK(t)
+
+			// Should return OK response
+			resp := req.Expect().
+				Status(http.StatusOK)
+			resp.chain.assertOK(t)
+
+			// Should set CheckRedirect
+			httpClient, _ := req.config.Client.(*http.Client)
+			assert.NotNil(t, httpClient.CheckRedirect)
+			assert.Equal(t, nil, httpClient.CheckRedirect(req.httpReq, nil))
+			assert.Equal(t, nil, httpClient.CheckRedirect(req.httpReq, make([]*http.Request, 1)))
+			assert.Equal(t,
+				errors.New("stopped after 1 redirects"),
+				httpClient.CheckRedirect(req.httpReq, make([]*http.Request, 2)),
+			)
+
+			// Should do round trip
+			assert.Equal(t, 2, tp.tripCount)
+		})
+
+		t.Run("request has no body and too many redirects", func(t *testing.T) {
+			tp := newMockTransportRedirect(t).
+				WithAssertFn(func(r *http.Request) bool {
+					assert.Equal(t, http.NoBody, r.Body)
+
+					if r.URL.String() == "/url" {
+						rc, err := r.GetBody()
+						assert.NoError(t, err)
+						assert.Equal(t, http.NoBody, rc)
+					} else if r.URL.String() == "/redirect" {
+						assert.Nil(t, r.GetBody)
+					} else {
+						t.Fatalf("invalid request URL")
+					}
+
+					return true
+				})
+
+			client := &http.Client{
+				Transport: tp,
+			}
+
+			config := Config{
+				Client:   client,
+				Reporter: reporter,
+			}
+
+			req := NewRequest(config, http.MethodPut, "/url").
+				WithRedirectPolicy(FollowAllRedirects).
+				WithMaxRedirects(1)
+			req.chain.assertOK(t)
+
+			// Should error
+			resp := req.Expect()
+			resp.chain.assertFailed(t)
+
+			// Should set CheckRedirect
+			httpClient, _ := req.config.Client.(*http.Client)
+			assert.NotNil(t, httpClient.CheckRedirect)
+			assert.Equal(t, nil, httpClient.CheckRedirect(req.httpReq, nil))
+			assert.Equal(t, nil, httpClient.CheckRedirect(req.httpReq, make([]*http.Request, 1)))
+			assert.Equal(t,
+				errors.New("stopped after 1 redirects"),
+				httpClient.CheckRedirect(req.httpReq, make([]*http.Request, 2)),
+			)
+
+			// Should do round trip
+			assert.Equal(t, 2, tp.tripCount)
+		})
+
+		t.Run("request has body", func(t *testing.T) {
+			tp := newMockTransportRedirect(t).
+				WithAssertFn(func(r *http.Request) bool {
+					assert.NotNil(t, r.Body)
+
+					if r.URL.String() == "/url" {
+						rc1, err := r.Body.(*bodyWrapper).GetBody()
+						assert.NoError(t, err)
+
+						rc2, err := r.GetBody()
+						assert.NoError(t, err)
+
+						assert.Equal(t, rc1, rc2)
+					} else if r.URL.String() == "/redirect" {
+						assert.Nil(t, r.GetBody)
+					} else {
+						t.Fatalf("invalid request URL")
+					}
+
+					return true
```

Comment:
Inside this WithAssertFn, we can just read r.Body and check that it's equal to "test body".

It's better not to assume that inside RoundTrip r.Body is bodyWrapper - because theoretically nothing prevents http.Client to wrap or replace body reader with something else when it passes body to RoundTripper.

Reply:
Same for similar tests below.